src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs

```diff
@@ -0,0 +1,487 @@
+webApp.MapGet("/customers/{customer}", (Customer customer) => {});
+
+public class Customer : IBindableFromHttpContext<Customer>
+{
+    public static async ValueTask<Customer?> BindAsync(HttpContext context, ParameterInfo parameter)
+    {
+        return new Customer();
+    }
+}
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Route_Parameter_withNullableType_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapGet("/customers/{customer}/contacts", (int? customer) => {});
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Handler_Parameter_withUnbindableComplexType_Fails()
+    {
+        // Arrange
+        var source = $$"""
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost(
+    "/customers",
+    ({|#0:[FromBody]Customer customer|}) => {
+    });
+
+public class Customer
+{
+}
+""";
+
+        var expectedDiagnostic = new DiagnosticResult(DiagnosticDescriptors.RouteHandlerParamterComplexTypeIsNotBindable)
+            .WithArguments("customer", "Customer")
+            .WithLocation(0);
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source, expectedDiagnostic);
+    }
+
+    [Fact]
+    public async Task Handler_Parameter_withInvalidBindAsyncImplementation_Fails()
+    {
+        // Arrange
+        var source = $$"""
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost(
+    "/customers",
+    ({|#0:[FromBody]Customer customer|}) => {
+    });
+
+public class Customer
+{
+    public async static Task<Customer> BindAsync(HttpContext context)
+    {
+        return new Customer();
+    }
+}
+""";
+
+        var expectedDiagnostic = new DiagnosticResult(DiagnosticDescriptors.RouteHandlerParamterComplexTypeIsNotBindable)
+            .WithArguments("customer", "Customer")
+            .WithLocation(0);
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source, expectedDiagnostic);
+    }
+
+    [Fact]
+    public async Task Handler_Parameter_withBindableComplexType_viaMethodConvention_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost(
+    "/customers",
+    ([FromBody]Customer customer) => {
+    });
+
+public class Customer
+{
+    public static async ValueTask<Customer?> BindAsync(HttpContext context)
+    {
+        return new Customer();
+    }
+}
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
```

Comment:
We probably should have an analyzer flag this case because we're going to try to deserialize `Customer` as JSON because of the `[FromBody]` attribute and that's probably not what people want if they're implementing `BindAsync`.

Reply:
I wonder how likely it is that folks might use the same type on different routes differently (JSON serialization works one way and they rely on ```BindAsync``` for another behavior). I guess they should implement a separate DTO in that situation.

Reply:
That's technically a possibility, and that flexibility one way our current behavior could be seen as "better". I don't think it's super realistic to want to support both JSON deserialization and custom `BindAsync` type, so I think this flexibility probably causes more confusion than it's worth.

I wouldn't patch the behavior because it's been that way since .NET 6, but I'd definitely consider changing it for 8.

Reply:
At the moment I've got it nope'ing out of anything for the ```[FromBody]``` scenario since the current behavior is that it will either work via JSON deserialization OR via BindAsync (if present). I think the throwing up an error/warning when BindAsync is not required, but does exist but doesn't conform to the expected interface is a bridge too far since its not like BindAsync is a reserved method name or anything - thoughts?

Reply:
> Also, what's the behavior when `IParsable<>` and/or `IBindableFromHttpContext<>` are implemented explicitly? Will this raise diagnostics about `TryParse` and/or `BindAsync` being incorrectly implmented? It probably should because we won't find them at runtime today, but it shouldn't as soon as #43546 gets fixed.

Do you think this is worth doing given that implementing interfaces explicitly is not that common and we actually want to change this behavior in the databinding plumbing anyway? I think this analyzer is going to ship with 8.0, I'm assuming that we'll have this explicit interface support in for 8.0?